File: app/src/main/java/com/example/spelltracker/Classes.java
package com.example.spelltracker;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Реестр D&D-классов приложения: id, отображаемое имя, файл со списком
 * заклинаний в assets/ и коэффициент участия в расчёте эффективного
 * уровня заклинателя.
 *
 * <p>Коэффициенты:
 * <ul>
 *   <li>{@code 1.0} — полный заклинатель (Бард, Волшебник, Друид, Жрец, Чародей);</li>
 *   <li>{@code 0.5} — половинный заклинатель (Паладин, Следопыт, Изобретатель);
 *       уровень делится на 2. Для Паладина и Следопыта — округление вниз
 *       ({@code roundUp=false}), для Изобретателя — округление вверх
 *       ({@code roundUp=true}, как в PHB);</li>
 *   <li>{@code 0.0} — не участвует в расчёте (Колдун использует pact magic).</li>
 * </ul>
 * </p>
 *
 * <p>Этот реестр — единственный источник правды для {@link ClassFilter}
 * (имена и списки заклинаний) и {@link SpellStorage} (расчёт уровня).</p>
 */
public final class Classes {

    public static final class Info {
        public final String id;
        public final String name;
        public final String assetFile;
        public final double factor;
        public final boolean roundUp;

        public Info(String id, String name, String assetFile, double factor, boolean roundUp) {
            this.id = id;
            this.name = name;
            this.assetFile = assetFile;
            this.factor = factor;
            this.roundUp = roundUp;
        }
    }

    public static final Info[] ALL = {
        new Info("bard",       "Бард",         "бард.json",         1.0, false),
        new Info("wizard",     "Волшебник",    "волшебник.json",    1.0, false),
        new Info("druid",      "Друид",        "друид.json",        1.0, false),
        new Info("cleric",     "Жрец",         "жрец.json",         1.0, false),
        new Info("warlock",    "Колдун",       "колдун.json",       0.0, false),
        new Info("paladin",    "Паладин",      "паладин.json",      0.5, false),
        new Info("ranger",     "Следопыт",     "следопыт.json",     0.5, false),
        new Info("sorcerer",   "Чародей",      "чародей.json",      1.0, false),
        new Info("artificer",  "Изобретатель", "изобретатель.json", 0.5, true),
    };

    private static final Map<String, Info> BY_ID;
    static {
        Map<String, Info> m = new LinkedHashMap<>();
        for (Info info : ALL) m.put(info.id, info);
        BY_ID = Collections.unmodifiableMap(m);
    }

    public static Map<String, Info> allById() {
        return BY_ID;
    }

    private Classes() {}
}

File: app/src/main/java/com/example/spelltracker/data/Classes.kt
package com.example.spelltracker.data

/**
 * Метаданные классов D&D 5e.
 *
 * Источник истины о том, какие классы вообще есть в приложении и как
 * они участвуют в подсчёте caster level по правилам мультикласса
 * из Player's Handbook:
 *
 *   factor = 1.0         — полный заклинатель (бард, волшебник, друид, жрец, чародей)
 *   factor = 0.5         — полузаклинатель (паладин, следопыт, изобретатель)
 *   factor = 0.0         — не заклинатель, но всё равно в сетке (колдун)
 *   roundUp = true       — для полузаклинателя округлять (lvl+1)/2
 *                          (нужно только изобретателю)
 *   isThirdCaster = true — Этап 19: 1/3 кастер (Eldritch Knight / Arcane
 *                          Trickster), caster level = ceil(lvl / 3).
 *                          Перебивает правило roundUp для дробных кастеров.
 *
 * Локализация имени класса: см. `com.example.spelltracker.ClassNames` —
 * `Info.nameRes()` возвращает `@StringRes Int`, UI резолвит через
 * `stringResource(...)`.
 *
 * Файл `assetFile` сохранён для справки (раньше загружали по одному
 * JSON на класс), но сейчас справочник заклинаний общий — один
 * `spells_normalized.json` в assets, см. [SpellParser.loadFromAssets].
 */
object Classes {

    data class Info(
        val id: String,
        @Suppress("unused") val assetFile: String,
        val factor: Double,
        val roundUp: Boolean,
        val isThirdCaster: Boolean = false,
    )

    val ALL: List<Info> = listOf(
        Info("bard",           "бард.json",         1.0, false),
        Info("wizard",         "волшебник.json",    1.0, false),
        Info("druid",          "друид.json",        1.0, false),
        Info("cleric",         "жрец.json",         1.0, false),
        Info("warlock",        "колдун.json",       0.0, false),
        Info("paladin",        "паладин.json",      0.5, false),
        Info("ranger",         "следопыт.json",     0.5, false),
        Info("sorcerer",       "чародей.json",      1.0, false),
        Info("artificer",      "изобретатель.json", 0.5, true),
        // Этап 19: архетипы-третьекастеры (1/3) — Воин/Eldritch Knight
        // и Плут/Arcane Trickster. По правилам пользователя caster level
        // считается как ceil(lvl / 3) — формула `(lvl + 2) / 3`. Это
        // **отличается** от стандартного PHB (там roundDown), но
        // соответствует тому, как игрок ведёт партию.
        Info("fighter_mystic", "fighter_mystic.json", 1.0 / 3.0, true, isThirdCaster = true),
        Info("rogue_mystic",   "rogue_mystic.json",   1.0 / 3.0, true, isThirdCaster = true),
    )

    val BY_ID: Map<String, Info> = ALL.associateBy { it.id }

    /**
     * Множество id классов, которые умеют колдовать (full/half/third caster).
     * Используется там, где нужно отличать «магов» от «не-магов» —
     * например, чтобы убрать Колдуна (factor=0) из списка заклинателей.
     */
    val CASTER_IDS: Set<String> = ALL.filter { it.factor > 0.0 }.map { it.id }.toSet()
}